lib/connectors/mongodb/connector.rb

```diff
@@ -56,30 +60,112 @@ def initialize(configuration: {})
         @direct_connection = configuration.dig(:direct_connection, :value)
       end
 
-      def yield_documents
+      def yield_documents(&on_doc_serialization)
+        check_filtering
+
+        call_db_function_on_collection(&on_doc_serialization)
+      end
+
+      private
+
+      def create_db_cursor_on_collection(collection)
+        return create_find_cursor(collection) if @advanced_filter_config[:find].present?
+
+        return create_aggregate_cursor(collection) if @advanced_filter_config[:aggregate].present?
+
+        collection.find
+      end
+
+      def check_filtering
+        return unless filtering_present?
+
+        check_find_and_aggregate
+      end
+
+      def check_find_and_aggregate
+        if @advanced_filter_config.keys.size != 1
+          invalid_keys_msg = "Only one of #{ALLOWED_TOP_LEVEL_FILTER_KEYS} is allowed in the filtering object. Keys present: '#{@advanced_filter_config.keys}'."
+          raise Utility::InvalidFilterConfigError.new(invalid_keys_msg)
+        end
+      end
+
+      def call_db_function_on_collection(&on_doc_serialization)
         with_client do |client|
           # We do paging using skip().limit() here to make Ruby recycle the memory for each page pulled from the server after it's not needed any more.
           # This gives us more control on the usage of the memory (we can adjust PAGE_SIZE constant for that to decrease max memory consumption).
           # It's done due to the fact that usage of .find.each leads to memory leaks or overuse of memory - the whole result set seems to stay in memory
           # during the sync. Sometimes (not 100% sure) it even leads to a real leak, when the memory for these objects is never recycled.
-          cursor = client[@collection].find
+          cursor, options = create_db_cursor_on_collection(client[@collection])
           skip = 0
 
+          found_overall = 0
+
+          # if no overall limit is specified by filtering use -1 to not break ingestion, when no overall limit is specified (found_overall is only increased,
+          # thus can never reach -1)
+          overall_limit = Float::INFINITY
+
+          if options.present?
+            # there could be a skip parameter defined for filtering
+            skip = options.fetch(:skip, skip)
+            # there could be a limit parameter defined for filtering -> used for an overall limit (not a page limit, which was introduced for memory optimization)
+            overall_limit = options.fetch(:limit, overall_limit)
+          end
+
+          overall_limit_reached = false
+
           loop do
-            found_count = 0
+            found_in_page = 0
+
             view = cursor.skip(skip).limit(PAGE_SIZE)
             view.each do |document|
-              yield serialize(document)
-              found_count += 1
+              yield serialize(document) { on_doc_serialization.call if block_given? }
```

Comment:
I _think_ this statement is redundant. I believe that you can remove all usages of `&on_doc_serialization` in parameter lists, and change this to:

```suggestion
              yield serialize(document)
```

Reply:
Actually, if I remove that statement 4 tests will fail (it seems like you cannot pass an implicit block two times? Or am I missing something obvious?). 

I think we need to pass the block explicitly to the `call_db_function_on_collection` to be able to `yield` it later. One other option would be to move all the code from `call_db_function_on_collection` directly into `yield_documents`. Then we can make use of an implicit block and we don't have to introduce the explicit block. I don't have a strong opinion here, but explicit blocks _always_ create a Proc object, which have a worse performance (source: [[Passing Blocks in Ruby Without &block](https://mudge.name/2011/01/26/passing-blocks-in-ruby-without-block/)]) We could use &Proc.new, but according to this article this option was removed in Ruby 3.0.0, so we probably run into the same issue again, if we use a version >= 3.0.0. 

WDYT?